File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.cpp
class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        
        vector<vector<int>> alp (26) ; 
        for(int i =0 ; i< s.size(); i++)
            alp[s[i]-'a'].push_back(i) ; 

        int ans = 0 ; 
        for(const auto& w: words){
            int x =-1; 
            bool have = true ; 

            for(char e:w){
                auto it = upper_bound( alp[e-'a'].begin() , alp[e-'a'].end() , x) ;
                if( it == alp[e-'a'].end())  have = false ; 
                else x = *it ; 
            }
            if(have) ans++ ; 
        }
        return ans ; 


        
        
        
    }
};

File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        
        dic = defaultdict(list)
        for i,e in enumerate(s): 
            dic[e].append(i) 

        ans = 0 
        for w in words: 

            ss = False ; cur = -1 ; i = 0 
            while True: 
                if w[i] in dic:
                    idxs = dic[w[i]]
                    have = False 
                    for ix in idxs: 
                        if ix > cur:
                            cur = ix
                            have = True 
                            break 
                    if not have: 
                        break 

                else: 
                    break 
                i +=1 
                if i >= len(w):
                    ss = True 
                    break 

            if ss: ans +=1
                
        return ans 
